src/librustdoc/html/static/js/search.js

```diff
@@ -1267,44 +1223,98 @@ function initSearch(rawSearchIndex) {
+             * One of them is actual unboxing (turning `Maybe a` into `a`), while the other is
+             * equivalent to converting a type parameter into an existential, which Hoogle doesn't
+             * seem to actually do? I notice that these two searches produce the same result:
+             * https://hoogle.haskell.org/?hoogle=Eq+-%3E+Set+Eq+-%3E+Set+Eq
+             * https://hoogle.haskell.org/?hoogle=Ord+-%3E+Set+Ord+-%3E+Set+Ord
+             *
+             * Haskell is a bit of a foreign language to me. I just want to be able to look up
+             * Option combinators without having to actually remember their (mostly arbitrary)
+             * names, and Haskell claims to already have that problem solved...
+             *
+             * @type Map<integer, FunctionType[]>
+             */
+            const unbox = function(fnType) {
+                if (fnType.id < 0) {
+                    const genid = (-fnType.id) - 1;
+                    if (whereClause && whereClause[genid]) {
+                        for (const trait of whereClause[genid]) {
+                            addFnTypeToFnTypeSet(trait);
+                        }
+                    }
+                    mgens.set(fnType.id, 0);
+                } else {
                     for (const innerFnType of fnType.generics) {
                         addFnTypeToFnTypeSet(innerFnType);
                     }
+                }
+            };
+            /**
+             * @param {FunctionType} fnType
+             */
+            const addFnTypeToFnTypeSet = function addFnTypeToFnTypeSet(fnType) {
+                const queryContainsArrayOrSliceElem = queryElemSet.has(typeNameIdOfArrayOrSlice);
+                // qsid is an index into the `queryElemSet`, while `fnType.id` is an index into
+                // the `fnTypeSet`. These are the same, except for generics, where they get mapped.
+                // If qsid = 0, it means the generic type parameter has undergone instance
+                // substitution, and
+                let qsid = fnType.id;
+                if (fnType.id < 0) {
+                    if (mgens.has(fnType.id)) {
+                        qsid = mgens.get(fnType.id);
+                    } else {
+                        qsid = null;
+                        searching: for (const qid of queryElemSet.keys()) {
+                            if (qid < 0) {
+                                for (const qidMapped of mgens.values()) {
+                                    if (qid === qidMapped) {
+                                        continue searching;
+                                    }
+                                }
+                                mgens.set(fnType.id, qid);
+                                qsid = qid;
+                                break;
+                            }
+                        }
+                    }
+                }
+                if (qsid === null || !(
```

Comment:
Took me a while to mentally parse this condition. Would it be possible to move it into a function and have conditions split to make reading easier? An inner function is completely fine.